AddedPlutusScriptAPI,  evalScripts, TxInfo, valContext etc.

TimSheard opened the pull request:
Added the module Cardano.Ledger.Alonzo.TxInfo
Had to add SafeToHash and HashAnnotated instances for Alonxo Tx.

This is an attempt to write the translation from Alonzo Tx to Plutus TxInfo.
There are many discontinuites between what the Ledger  understands as a Tx and the information that Plutus wants included in TxInfo.  This start isolates some of those differences, and is meant to start a discussion about what shoud really be in TxInfo on the Plutus side.

The next step, nce that is straightened out, is to embed TxInfo into Plutus Data.

TimSheard committed:
WIP began writing transTx:: Tx era -> Plutus.TxInfo

Added the module Cardano.Ledger.Alonzo.TxInfo
Had to add SafeToHash and HashAnnotated instances for Alonxo Tx.

TimSheard committed:
Added the file TxInfo.hs

TimSheard committed:
ormolise

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+transTxId :: TxId era -> P.TxId
+transTxId (TxId safe) = P.TxId (transSafeHash safe)
+
+transVI :: ValidityInterval -> P.SlotRange
```
I'm on the fence about whether we should just switch to a simpler interval type like yours. Ours is a bit more complex to make it easier for people to write the interval they actually want with the endpoint properties they actually want, and then have intersecting them etc. do the right thing.

TimSheard replied:
up to you, I'm fine with it like this.

michaelpj commented:
@j-mueller can you have a look at this too.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+transTxIn txin =
+  P.TxInInfo
+    (transTxIn' txin)
+    Nothing -- NOT SURE ABOUT THIS
```
Yeah, not sure what to do here. The idea is for this to be a "resolved" transaction input, so we'd like to decorate it with the information about the output that's being spent, including the value on it, the address etc.

So we probably want something like
```
data TxInInfo {
   outRef :: TxOutRef,
   someKindOfAddressInformation :: ???,
   value :: Value
}
```
Or maybe even:
```
data TxInInfo {
  outRef :: TxOutRef,
  out :: TxOut
}
```

A final question is whether we directly attach the redeemer hash. In the real transaction the redeemers are in a separate structure, but a) there must be a redeemer for every input otherwise we won't even get to this stage, and b) there's no virtue in making people look it up. So I vote for putting the redeemer hash in here too.

TimSheard replied:
went with the latter.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+  P.TxInInfo
+    (transTxIn' txin)
+    Nothing -- NOT SURE ABOUT THIS
+    (transValue (inject @(Mary.Value c) (Coin 0))) -- OR ABOUT THIS
```
This needs to be the value on the spent output - this is the place where we need the UTXO set!

TimSheard replied:
done

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+
+transTxOut :: forall era. (Era era, Value era ~ Mary.Value (Crypto era)) => TxOut era -> P.TxOut
+transTxOut (TxOut (Addr _network (ScriptHashObj sh) _stake) val datahash) =
+  P.TxOut
```
Nit: can we use record constructor syntax more in this file? My poor brain can't remember the ordering of all the fields...

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    (transValue (inject @(Mary.Value c) (Coin 0))) -- OR ABOUT THIS
+
+transTxOut :: forall era. (Era era, Value era ~ Mary.Value (Crypto era)) => TxOut era -> P.TxOut
+transTxOut (TxOut (Addr _network (ScriptHashObj sh) _stake) val datahash) =
```
We definitely want to use `_stake` in due course! I'm pretty sure we should ignore `_network`, though.

michaelpj replied:
In doing so, I think we should follow the structure you've got and have an `Addr` type with a field which is a script address or a pubkey address.

TimSheard replied:
We ignore the network and introduce a new type StakeingHash

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+  P.TxOut
+    (P.PubKeyAddress (transKeyHash kh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
```
Well this can't be right: this isn't a script hash address, why do we have a data hash?

(Edit: ah, this is just going to trigger line 125, I guess)

TimSheard replied:
yes that is right

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    (P.PubKeyAddress (transKeyHash kh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+transTxOut (TxOut (AddrBootstrap (BootstrapAddress _ad)) val datahash) =
```
What's a bootstrap address? Do we care about scripts seeing them? (sounds like the sort of thing that may not be very common!)

TimSheard replied:
They have been elided from the TxInfo

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+  P.TxOut
+    (P.ScriptAddress (transScriptHash sh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
```
This `ScriptType` type is exactly isomorphic to your `Maybe DataHash` type: I say we kill it and just use `Maybe` on our side too

TimSheard replied:
no longer applies

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+transScriptHash :: ScriptHash i -> P.ValidatorHash
+transScriptHash (ScriptHash (UnsafeHash h)) = P.ValidatorHash (fromShort h)
+
+transDataHash :: StrictMaybe (DataHash c) -> P.DatumHash
```
I think we should follow your naming for `KeyHash`, `ScriptHash`... but I think it should be `DatumHash` on your side?

TimSheard replied:
left it for now

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+  P.TxInfo
+transTx tx =
+  P.TxInfo
+    (map transTxIn (Set.toList allinputs))
```
Possible non-determinism alert! I guess it's probably fine... do we guarantee that the `Ord` instance on inputs is stable?

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    (transValue (inject @(Mary.Value (Crypto era)) fee))
+    (transValue forge)
+    (transVI interval)
+    [] -- MonetaryPolicyHash ???
```
bin it - this is from when we thought people were going to have to list the set of minting policies explicitly.

TimSheard replied:
removed

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    (transValue forge)
+    (transVI interval)
+    [] -- MonetaryPolicyHash ???
+    [] -- PubKeyHash ???
```
Should be the new `RequiredSigners` field

TimSheard replied:
done

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    [] -- MonetaryPolicyHash ???
+    [] -- PubKeyHash ???
+    [] -- Auxiliary Data and hashes
+    (P.TxId (transSafeHash (hashAnnotated @(Crypto era) tx)))
```
`transTxId`?

TimSheard replied:
we want the hash of the TxBody. That is now fixed.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,155 @@
+    _witnesses = wits' tx
+    _isval = isValidating' tx
+    _auxdat = auxiliaryData' tx
+    allinputs = Set.union (inputs' tbody) (inputs_fee' tbody)
```
I think we decided we'd list them separately?

TimSheard replied:
Split  into two fields.

michaelpj reviewed (commented):
Looks good so far!

TimSheard committed:
Added transValue, imported Plutus data and the IsData class.

TimSheard committed:
ormolise

TimSheard committed:
Added module RunPlutus, wrote function valContext.

TimSheard committed:
Added module Cardano.Ledger.Alonzo.FakePlutus

This module lets us experiment with small changes to the Plutus TxINfo and associated types.
We made a few changes to make the structure more closely align with the actual Ledger code.
Got close to completion on translating a Alonzo.Tx into a FakePlutus.TxInfo. Next step
mke the changes in FakePlutus real in Plutus.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/FakePlutus.hs:
```diff
@@ -0,0 +1,73 @@
+  { txInInfoOutRef :: !TxOutRef,
+    txInAddr :: !Address,
+    txInValue :: !Value,
+    txInDataHash :: !(Maybe DatumHash)
```
Why inline the `TxOut` fields instead of just having a field of type `TxOut`? I guess there's a usability/maintainability tradeoff here... my instinct is to go for `data TxInfo = TxInInfo { txInInfoRef :: TxOutRef, txInInfoOut :: TxOut }`

michaelpj replied:
I'd also be tempted to have a `Maybe RedeemerHash` here. We need to decide if we're doing the pointer thing or what.

TimSheard replied:
We now use TxOut, rather than inlining

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/FakePlutus.hs:
```diff
@@ -0,0 +1,73 @@
+    txInfoSignatories :: [PubKeyHash],
+    txInfoData :: [(DatumHash, Datum)],
+    -- | Hash of the pending transaction (excluding witnesses)
+    txInfoId :: TxId
```
Also need withdrawals and certificate registrations, and we need to do *something* about the redeemers.

JaredCorduan replied:
you probably want _all_ the certificates, right? delegations, de-registrations, etc. maybe even the governance ones?

michaelpj replied:
I guess so? I don't know what these things are, but if we think people will want to make assertions about them then we need them. I guess they'll mostly just not be present, though.

JaredCorduan replied:
I think this is pretty sensible:

do include:
* stake credential registration and deregistration
* delegation to a stake pool
* pool registration / re-registration
* pool retirement

do not include:
* genesis key delegations
* MIR certs (the mechanism that allows the genesis delegates to move lovelace out of the reserves and treasury)

TimSheard replied:
We now have extra fields for all these things.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -22,56 +35,82 @@ import Cardano.Ledger.Alonzo.TxBody
+transStakeReference _mp (StakeRefBase cred) = Just (P.StakingHash (transStakeCred cred))
+transStakeReference mp (StakeRefPtr ptr) =
+  case Map.lookup ptr mp of
+    Nothing -> Nothing
```
Is this impossible if we've got to this stage of validation? Does the ledger allow addresses with dangling staking references? If it's impossible, could we make this fail validation instead in a more aggressive way? I'm not sure it will matter in practice, it just feels itchy to not be able to see the difference between this invariant failing and plain old `StakeRefNull`.

JaredCorduan replied:
>  Does the ledger allow addresses with dangling staking references?

yes, it does. someone can place lovelace in a pointer address and then de-register their stake credential, leaving a dangling staking references. so we don't even bother to enforce valid pointers on creation of an output. Addresses with dangling pointers behave identically like `StakeRefNull`. But if plutus would like to make such a distinction, we could pass some `Dangling` type to the interpreter.

michaelpj replied:
Hmmmmmm. I don't *think* so. I'm trying to think of edge cases, but I think it's fine. Worth a comment, perhaps.

JaredCorduan replied:
sure, comments are always great. but I'm honestly happy either way. we keep it as is unless you think of something.

JaredCorduan replied:
we discussed this in person, but I'll make a note here. we are _not_ going to resolve pointer addresses. that would break determinism.

TimSheard replied:
We now pas in pointers, but we do not resolve them, so you can check if two pointers are the same, but you can't tell what they point to.

michaelpj commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -196,20 +225,32 @@ transTx utxo tx =
     fee = txfee' tbody
     forge = mint' tbody
     interval = vldt' tbody
+    vkkeyset = txwitsVKey' (wits' tx)
```
This needs a scary FIXME: whatever we do it can't be this (see discussion about signature stripping)

JaredCorduan replied:
@TimSheard we can pass plutus the `reqSignerHashes` field of the transaction once #2198 is merged.

TimSheard replied:
It is now fixed.

TimSheard committed:
Added to FakePlutus

DCert, Wdrls, and others
Worked out how to actually call plutus in eval_scripts.

TimSheard committed:
Renamed RunPlutus, completed ValidateScript, evalScripts.

Renamed module Caradano.Ledger.Alonzo.RunPlutus to Caradano.Ledger.Alonzo.PlutusScriptApi.
Completed the ValidateScript instance for Alonzo, and the evalScripts function.
Cleaned up the Utxos modue, that was way over constrained.

TimSheard committed:
Cleaned up, Added comments

Renamed NativeScript constructor of Script to TimelockScript.

TimSheard committed:
Added the type Context and ScriptPurpose to FakePlutus.

TimSheard commented:
We are past the WIP stage.  I think we should merge, and workout the FakePlutus later.

redxaxder commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -0,0 +1,222 @@
+   in [ (script, d : (valContext txinfo sp ++ getData tx utxo sp), eu, cost)
+        | (sp, scripthash) <- scriptsNeeded utxo tx, -- TODO, IN specification ORDER IS WRONG
+          (d, eu) <- maybeToList (indexedRdmrs tx sp),
+          script <- -- onlytwoPhaseScripts tx scripthash
```
you meant to remove this

TimSheard replied:
It is now removed.

redxaxder commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -0,0 +1,222 @@
+      isNativeScript @era s
+        || ( isJust (indexedRdmrs tx sp)
+               && (not (isSpending sp) || not (null (getData tx utxo sp)))
+           )
```
this might be clearer written in terms of `needsData` and `hasData` predicates, and checking that if we need the data then we have it.

TimSheard replied:
done

redxaxder commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -0,0 +1,222 @@
+    !reward = foldl' accum [] (Map.keys m2)
+      where
+        (Wdrl m2) = getField @"wdrls" txb
+        accum !ans !accnt = case getRwdCred accnt of -- TODO  IS THIS RIGHT?
```
is this right?

TimSheard replied:
removed comment

redxaxder commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,337 @@
+  UTxO era ->
+  TxIn (Crypto era) ->
+  [P.TxInInfo] ->
+  [P.TxInInfo]
```
this choice of types puts greater demands on the reader's knowledge than most alternatives

TimSheard replied:
Both now return a Maybe type.

JaredCorduan reviewed (approved):
other than the few things we've discussed on the call that we still want to tweak, looks good to me!

TimSheard committed:
Addressed comments by Alex, Jared, and Polina
